### PR TITLE
Fixes to reading and rewriting

### PR DESCRIPTION
**Pull request by djspiewak**
Collapse wasn't working, so I had to add a `Unary` node which is rewritten away immediately.  Also `LPtoQS` wasn't working at all due to `\circ` precedence.  All that is fixed.  We're very close to graduating.  Just have to get that thesis banged out…

### CODE REVIEW

**Comment by alissapajer** [comment]
I'd like to see exactly what goes wrong if we use `Map` in place of `Unary` using the new version of `rewriteM`. Any problem with `FreeMap` equality will crop up again, so we should understand that more deeply now before masking over it with `Unary`.

**Comment by djspiewak** [comment]
@alissapajer It's not just about `Equal[FreeMap]`, though that does seem to be problematic.  The bigger issue is that there is no sane `Order[FreeMap]`, much less `Hash[FreeMap]`, which is what I *really* need to do this.  The compaction algorithm is fundamentally dependent on an `SMap[QSU, Symbol]`, which naturally requires equality and hash checks on `QSU`.  Any other data structure there destroys the asymptotic complexity of the algorithm (e.g. `EqMap` would be O(n^2)).

**Comment by alissapajer** [comment]
@djspiewak can you add some scaladoc about why have `Unary`then? A sentence saying we use `SMap[QSU, Symbol]` so we need sane comparison checks on `QSU` is all we need. We currently have one giant blob of `QSU` types, so it's impossible to tell just by looking at them which are used in which contexts. `Unary` appears unnecessary unless you know it's being used in a comparison situation.

**Comment by alissapajer** [comment]
Down the road when we're figuring out how to break apart these various `QSU` types, hashable-ness will be a useful distinction.

**Comment by djspiewak** [comment]
@alissapajer Willdo.  I meant to document it, but never got back round to it.

File: connector/src/main/scala/quasar/qscript/qsu/QSUGraph.scala [hunk 1]
```diff
@@ -90,17 +90,41 @@ final case class QSUGraph[T[_[_]]](
       for {
         visited <- MS.get
 
-        back <- if (visited(root)) {
+        back <- if (visited(g.root)) {
           this.point[G]
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/QSUGraph.scala]: This should be `g.point[G]`, no?

Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/QSUGraph.scala]: 👍 🤦‍♂️

**Pull request merged**